content/blog/provisioned-concurrency-insurance-against-cold-starts-in-aws-lambda/index.md

```diff
@@ -0,0 +1,154 @@
+
+A trick known as **Lambda Warmers** uses kept-alive workers to reduce the frequency of cold starts. The idea is to have a CloudWatch timer that fires every few minutes and sends `N` parallel requests to the target Lambda. If all those requests land at the same time, AWS has to provision at least `N` workers to process them. The actual operation doesn't have to do any useful work, but it resets the recycling timer back to zero.
+
+There are a few drawbacks with this approach:
+
+* If a valid request comes at the same time as warming requests, it might hit a cold start.
+* Extra logic is needed in the Lambda code to detect warming requests and short-circuit instead of doing useful work.
+* CloudWatch rules require additional setup and management.
+* The result is still best-effort: Lambda still recycles workers from time to time.
+
+You can find more details about how Pulumi can help with some of these issues in [AWS Lambda Warmer as Pulumi Component](https://mikhail.io/2018/08/aws-lambda-warmer-as-pulumi-component/).
+
+Here is an example of configuring the provisioned concurrency with Pulumi in TypeScript:
+
+```ts
+const lambda = new aws.lambda.Function("mylambda", {
+    code: new pulumi.asset.FileArchive("./app"),
+    handler: "handler",
+    role: role.arn,
+    publish: true,
+});
+
+const fixed = new aws.lambda.ProvisionedConcurrencyConfig("fixed-concurrency", {
+  functionName: lambda.name,
+  qualifier: lambda.version,
+  provisionedConcurrentExecutions: 2,
+});
+```
+
+The snippet sets the provisioned concurrency for `mylambda` to a fixed value of `2`. Note that concurrency is provisioned per Lambda version, and it can't be set for the `$LATEST` version alias. Therefore, I set the `publish` attribute of my Lambda to `true` and reference a specific version of the Lambda in provisioning.
+
+## Dynamic Provisioned Concurrency
+
+A fixed provisioned concurrency works good for a stable workloads.
+
+{{< figure src="./steady.png" caption="Fixed provisioned concurrency for uniform workloads" >}}
+
+However, many workloads fluctuate a lot. Extreme elasticity and lack of configuration parameters have always been the essential benefits of AWS Lambda. It works great if you can tolerate the cold starts that come during scale-out. If not, you can explore more advanced scenarios of provisioning concurrency.
+Instead of choosing a permanently fixed value, you can configure provisioned concurrency to autoscale. The first required component is the autoscaling target:
+
+```ts
+const resourceId = pulumi.interpolate`function:${lambda.name}:${lambda.version}`;
+const target = new aws.appautoscaling.Target("target", {
+    resourceId,
+    serviceNamespace: "lambda",
+    scalableDimension: "lambda:function:ProvisionedConcurrency",
+    minCapacity: 1,
+    maxCapacity: 10,
+});
+```
+
+The second component is the autoscaling policy, which defines the conditions when scaling starts. There are two important way
```

Comment:
Incomplete sentence and/or section?

Reply:
Indeed... restoring as well, and checking the rest of the document :(

Reply:
I'm not sure what happened here. A lot of the edits I made are no longer there.